elasticapm/contrib/sanic/__init__.py

```diff
@@ -0,0 +1,338 @@
+
+    >>> apm = ElasticAPM(app=app)
+
+    Configure the APM Client Using Custom Configurations::
+
+    >>> apm = ElasticAPM(app=app, config={
+        "SERVICE_NAME": "elastic-apm-sample",
+        "SERVICE_VERSION": "v1.2.0",
+        "SERVER_URL": "http://eapm-server.somdomain.com:443",
+        "SECRET_TOKEN": "supersecrettokenstuff",
+    })
+
+    Pass a pre-build Clinet instance to the APM Middleware::
+
+    >>> apm = ElasticAPM(app=app, client=Client())
+
+    Pass arbitrary Server name and token to the client while initializing::
+
+    >>> apm = ElasticAPM(app=app, service_name="elastic-apm-sample", secret_token="supersecretthing")
+
+    Capture an Exception::
+
+    >>> try:
+    >>>     1 / 0
+    >>> except ZeroDivisionError:
+    >>>     apm.capture_exception()
+
+    Capture generic message::
+
+    >>> apm.capture_message("Some Nice message to be captured")
+    """
+
+    def __init__(
+        self,
+        app: Sanic,
+        client: t.Optional[Client] = None,
+        client_cls: t.Type[Client] = Client,
+        config: APMConfigType = None,
+        transaction_name_callback: TransactionNameCallbackType = None,
+        user_context_callback: UserInfoCallbackType = None,
+        custom_context_callback: CustomContextCallbackType = None,
+        label_info_callback: LabelInfoCallbackType = None,
+        **defaults,
+    ) -> None:
+        """
+        Initialize an instance of the ElasticAPM client that will be used to configure the reset of the Application
+        middleware
+
+        :param app: An instance of Sanic app server
+        :param client: An instance of Client if you want to leverage a custom APM client instance pre-created
+        :param client_cls: Base Instance of the Elastic Client to be used to setup the APM Middleware
+        :param config: Configuration values to be used for setting up the Elastic Client. This includes the APM server
+        :param transaction_name_callback: Callback method used to extract the transaction name. If nothing is provided
+            it will fallback to the default implementation provided by the middleware extension
+        :param user_context_callback: Callback method used to extract the user context information. Will be ignored
+            if one is not provided by the users while creating an instance of the ElasticAPM client
+        :param custom_context_callback: Callback method used to generate custom context information for the transaction
+        :param label_info_callback: Callback method used to generate custom labels/tags for the current transaction
+        :param defaults: Default configuration values to be used for settings up the APM client
+        """
+        self._app = app  # type: Sanic
+        self._client_cls = client_cls  # type: type
+        self._client = client  # type: t.Union[None, Client]
+        self._skip_headers = defaults.pop("skip_headers", [])  # type: t.List[str]
```

Comment:
I'm not a huge fan of introducing framework-specific configuration that doesn't need to be framework specific.

We do already have a way to sanitize HTTP headers in a framework agnostic way:

https://www.elastic.co/guide/en/apm/agent/python/current/sanitizing-data.html#sanitizing-data

Since 6.0, the list of names that trigger sanitization is [configurable](https://www.elastic.co/guide/en/apm/agent/python/current/configuration.html#config-sanitize-field-names).

Reply:
Makes sense. Let me get rid of this custom one and fall back to using the default support that is provided by the agent framework. Thanks for the heads up on this.